### ISSUE TITLE
[STORM-969] HDFS Bolt can end up in an unrecoverable state

### ISSUE THREAD

**Issue by jira-importer** [apache/storm#4807]
<p>The body of the HDFSBolt.execute() method is essentially one try-catch block.  The catch block reports the error and fails the current tuple.  In some cases the bolt's FSDataOutputStream object (named 'out') is in an unrecoverable state and no subsequent calls to execute() can succeed.</p>

<p>To produce this scenario:</p>
<ul class="alternate" type="square">
	<li>process some tuples through HDFS bolt</li>
	<li>put the underlying HDFS system into safemode</li>
	<li>process some more tuples and receive a correct ClosedChannelException</li>
	<li>take the underlying HDFS system out of safemode</li>
	<li>subsequent tuples continue to fail with the same exception</li>
</ul>

<p>The three fundamental operations that execute takes (writing, sync'ing, rotating) need to be isolated so that errors from each are specifically handled.</p>

---
<details><summary><i>Originally reported by <a title="Aaron Dossett" href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dossett@gmail.com">dossett@gmail.com</a>, imported from: <a href="https://issues.apache.org/jira/browse/STORM-969" target="_blank">HDFS Bolt can end up in an unrecoverable state</a></i></summary>
<i><ul>
<li><b>assignee</b>: <a title="Aaron Blake Niskode-Dossett" href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dossett">dossett</a>
<li><b>status</b>: Resolved
<li><b>priority</b>: Major
<li><b>resolution</b>: Fixed
<li><b>resolved</b>: 2015-10-05T15:23:37+01:00
<li><b>imported</b>: 2025-01-24
</ul></i>
</details>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dossett@gmail.com">dossett@gmail.com</a>:</i>
<p>I should have a PR for this shortly.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>GitHub user dossett opened a pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/664" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/664</a></p>

<p>    <a href="https://issues.apache.org/jira/browse/STORM-969" title="HDFS Bolt can end up in an unrecoverable state" class="issue-link" data-issue-key="STORM-969"><del>STORM-969</del></a>: HDFS Bolt can end up in an unrecoverable state</p>

<p>    A few notes about this PR:</p>
<ul class="alternate" type="square">
	<li>I updated the storm-hdfs pom.xml to align with other external modules.  Most significant change was probably going from hdfs version 2.2 to ${hadoop.version} (i.e. currently 2.6)</li>
</ul>


<ul class="alternate" type="square">
	<li>Many errors are recovered by forcing a file rotation which opens a new, valid File.  So the rotation now occurs either according to rotation policy or when a serious error happens.  Work could probably be done to reopen the same file name to reduce the number of rotations.</li>
</ul>


<ul class="alternate" type="square">
	<li>Added unittests with MiniDFSCluster</li>
</ul>


<p>You can merge this pull request into a Git repository by running:</p>

<p>    $ git pull <a href="https://github.com/dossett/storm" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/dossett/storm</a> <a href="https://issues.apache.org/jira/browse/STORM-969" title="HDFS Bolt can end up in an unrecoverable state" class="issue-link" data-issue-key="STORM-969"><del>STORM-969</del></a></p>

<p>Alternatively you can review and apply these changes as the patch at:</p>

<p>    <a href="https://github.com/apache/storm/pull/664.patch" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/664.patch</a></p>

<p>To close this pull request, make a commit to your master/trunk branch<br/>
with (at least) the following in the commit message:</p>

<p>    This closes #664</p>

<hr />
<p>commit 795aaf93af78bf664727b91c179e0d96f673f674<br/>
Author: Aaron Dossett <aaron.dossett@target.com><br/>
Date:   2015-08-02T22:22:51Z</p>

<p>    <a href="https://issues.apache.org/jira/browse/STORM-969" title="HDFS Bolt can end up in an unrecoverable state" class="issue-link" data-issue-key="STORM-969"><del>STORM-969</del></a>: HDFS Bolt can end up in an unrecoverable state</p>

<hr />

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user harshach commented on the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/664#issuecomment-131588005" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/664#issuecomment-131588005</a></p>

<p>    @arunmahadevan since you've made changes to HdfsBolt as part of <a href="https://issues.apache.org/jira/browse/STORM-837" title="HdfsState ignores commits" class="issue-link" data-issue-key="STORM-837"><del>STORM-837</del></a> can you take look at this PR check if its already handled by your changes.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user arunmahadevan commented on the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/664#issuecomment-131602837" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/664#issuecomment-131602837</a></p>

<p>    @harshach the changes I made were in the trident implementation (HDFSState) which is independent of this. Anyways I reviewed the changes.</p>

<p>    Overall it appears that the exceptions are now handled individually at write, sync and rotate phases. But I see a few issues with the change.</p>
<ul class="alternate" type="square">
	<li>the tuples are acked only on sync - if the tuple rates are low, they will never be acked and keep timing out and same tuples will be replayed again and again.</li>
	<li>there is an attempt to sync even when the write fails with an IOException. Since write already threw an IOException, chances are high that the sync would also fail with IOException.</li>
</ul>


<p>    I think it might be simpler to keep the existing logic and just rotate the file whenever we see an IOException (or maybe after a few times we repeatedly hit the IOException) and completely fail by propagating the exception up if the situation does not improve after a few rotations.</p>

<p>    Also I see that the existing implementation acks the tuples before actually syncing them to disk, which might result in data loss. I think we should change this to ack only after an hsync and have a sync policy that considers both count and time based thresholds.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user dossett commented on the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/664#issuecomment-131648086" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/664#issuecomment-131648086</a></p>

<p>    @arunmahadevan Thank you for the feedback!  I have added a tick tuple feature to address your first point (I am already using this locally, I forgot to include it in this PR).</p>

<p>    You are right that I attempt to sync even when the write fails.  I agree that this is unlikely to succeed, but it seems worthwhile to try to "save" as many tuples as possible.</p>

<p>    With this tick tuple change I think the code is very close to your recommendation.  If the bolt is in a bad state the tick tuples will periodically try to rotate the file and all errors will be reported up through ```this.collector.reportError(e)```</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user arunmahadevan commented on a diff in the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/664#discussion_r37160568" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/664#discussion_r37160568</a></p>

<p>    &#8212; Diff: external/storm-hdfs/src/main/java/org/apache/storm/hdfs/bolt/HdfsBolt.java &#8212;<br/>
    @@ -80,6 +86,11 @@ public HdfsBolt addRotationAction(RotationAction action)</p>
{
     return this;
 }

<p>    +    public HdfsBolt withTickTupleIntervalSeconds(int interval) {<br/>
    &#8212; End diff &#8211;</p>

<p>    Could give a more meaningful name to convey the actual usage (e.g `withFlushIntervalSeconds`). I think we need to have a default value for this to ensure a sync is `always` done even if user doesn't specify this option and also ensure the option value is within some lower and upper thresholds so that tuples are acked within TOPOLOGY_MESSAGE_TIMEOUT_SECS and a sync doesn't happen too frequently.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user arunmahadevan commented on a diff in the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/664#discussion_r37160668" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/664#discussion_r37160668</a></p>

<p>    &#8212; Diff: external/storm-hdfs/src/main/java/org/apache/storm/hdfs/bolt/HdfsBolt.java &#8212;<br/>
    @@ -88,35 +99,94 @@ public void doPrepare(Map conf, TopologyContext topologyContext, OutputCollector</p>

<p> @Override<br/>
 public void execute(Tuple tuple) {</p>
<ul class="alternate" type="square">
	<li>try {</li>
	<li>byte[] bytes = this.format.format(tuple);</li>
	<li>synchronized (this.writeLock) {</li>
	<li>out.write(bytes);</li>
	<li>this.offset += bytes.length;<br/>
    -</li>
	<li>if (this.syncPolicy.mark(tuple, this.offset)) {</li>
	<li>if (this.out instanceof HdfsDataOutputStream) 
{
    -((HdfsDataOutputStream) this.out).hsync(EnumSet.of(SyncFlag.UPDATE_LENGTH));
    -    }
<p> else </p>
{
    -this.out.hsync();
    -    }</li>
	<li>this.syncPolicy.reset();<br/>
    +boolean forceRotate = false;<br/>
    +synchronized (this.writeLock) {<br/>
    +    boolean forceSync = false;<br/>
    +    if (TupleUtils.isTick(tuple)) 
{
    +LOG.debug("TICK! forcing a file system flush");
    +forceSync = true;
    +    }
<p>    +    else </p>
<div class="error"><span class="error">Unknown macro: {    +try {
    +    writeAndAddTuple(tuple);
    +} catch (IOException e) {
    +    //If the write failed, try to sync anything already written
    +    LOG.info("Tuple failed to write, forcing a flush of existing data.");
    +    this.collector.reportError(e);
    +    forceSync = true;
    +    this.collector.fail(tuple);
     } }</span> </div></li>
</ul>


<ul class="alternate" type="square">
	<li>this.collector.ack(tuple);<br/>
    +    if (this.syncPolicy.mark(tuple, this.offset) || forceSync) 
<div class="error"><span class="error">Unknown macro: {    +try {
    +    syncAndAckTuples();
    +} catch (IOException e) {
    +    LOG.warn("Data could not be synced to filesystem, failing this batch of tuples");
    +    this.collector.reportError(e);
    +    //Force rotation to get a new file handle
    +    forceRotate = true;
    +    for (Tuple t : tupleBatch)
    +this.collector.fail(t);
    +    tupleBatch.clear();
    +}    +    }</span> </div>
<p>    +}</p></li>
</ul>


<ul class="alternate" type="square">
	<li>if(this.rotationPolicy.mark(tuple, this.offset)){</li>
	<li>rotateOutputFile(); // synchronized</li>
	<li>this.offset = 0;</li>
	<li>this.rotationPolicy.reset();<br/>
    +if(this.rotationPolicy.mark(tuple, this.offset) || forceRotate) {<br/>
    +    try 
{
    +rotateAndReset();
    +    }
<p> catch (IOException e) </p>
{
    +this.collector.reportError(e);
    +LOG.warn("File could not be rotated");
    +//At this point there is nothing to do.  In all likelihood any filesystem operations will fail.
    +//The next tuple will almost certainly fail to write and/or sync, which force a rotation.  That
    +//will give rotateAndReset() a chance to work which includes creating a fresh file handle.
 }</li>
	<li>} catch (IOException e) 
{
    -    this.collector.reportError(e);
    -    this.collector.fail(tuple);
     }
<p> }</p></li>
</ul>


<p>    +    private void rotateAndReset() throws IOException </p>
{
    +rotateOutputFile(); // synchronized
    +this.offset = 0;
    +this.rotationPolicy.reset();
    +    }
<p>    +<br/>
    +    private void syncAndAckTuples() throws IOException {<br/>
    &#8212; End diff &#8211;</p>

<p>    An optimization could be to do to the sync only if some tuples were actually written out since last sync.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user arunmahadevan commented on a diff in the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/664#discussion_r37161325" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/664#discussion_r37161325</a></p>

<p>    &#8212; Diff: external/storm-hdfs/src/main/java/org/apache/storm/hdfs/bolt/HdfsBolt.java &#8212;<br/>
    @@ -88,35 +99,94 @@ public void doPrepare(Map conf, TopologyContext topologyContext, OutputCollector</p>

<p> @Override<br/>
 public void execute(Tuple tuple) {</p>
<ul class="alternate" type="square">
	<li>try {</li>
	<li>byte[] bytes = this.format.format(tuple);</li>
	<li>synchronized (this.writeLock) {</li>
	<li>out.write(bytes);</li>
	<li>this.offset += bytes.length;<br/>
    -</li>
	<li>if (this.syncPolicy.mark(tuple, this.offset)) {</li>
	<li>if (this.out instanceof HdfsDataOutputStream) 
{
    -((HdfsDataOutputStream) this.out).hsync(EnumSet.of(SyncFlag.UPDATE_LENGTH));
    -    }
<p> else </p>
{
    -this.out.hsync();
    -    }</li>
	<li>this.syncPolicy.reset();<br/>
    +boolean forceRotate = false;<br/>
    +synchronized (this.writeLock) {<br/>
    +    boolean forceSync = false;<br/>
    +    if (TupleUtils.isTick(tuple)) 
{
    +LOG.debug("TICK! forcing a file system flush");
    +forceSync = true;
    +    }
<p>    +    else </p>
<div class="error"><span class="error">Unknown macro: {    +try {
    +    writeAndAddTuple(tuple);
    +} catch (IOException e) {
    +    //If the write failed, try to sync anything already written
    +    LOG.info("Tuple failed to write, forcing a flush of existing data.");
    +    this.collector.reportError(e);
    +    forceSync = true;
    +    this.collector.fail(tuple);
     } }</span> </div></li>
</ul>


<ul class="alternate" type="square">
	<li>this.collector.ack(tuple);<br/>
    +    if (this.syncPolicy.mark(tuple, this.offset) || forceSync) 
<div class="error"><span class="error">Unknown macro: {    +try {
    +    syncAndAckTuples();
    +} catch (IOException e) {
    +    LOG.warn("Data could not be synced to filesystem, failing this batch of tuples");
    +    this.collector.reportError(e);
    +    //Force rotation to get a new file handle
    +    forceRotate = true;
    +    for (Tuple t : tupleBatch)
    +this.collector.fail(t);
    +    tupleBatch.clear();
    +}    +    }</span> </div>
<p>    +}</p></li>
</ul>


<ul class="alternate" type="square">
	<li>if(this.rotationPolicy.mark(tuple, this.offset)){</li>
	<li>rotateOutputFile(); // synchronized</li>
	<li>this.offset = 0;</li>
	<li>this.rotationPolicy.reset();<br/>
    +if(this.rotationPolicy.mark(tuple, this.offset) || forceRotate) {<br/>
    +    try 
{
    +rotateAndReset();
    +    }
<p> catch (IOException e) </p>
{
    +this.collector.reportError(e);
    +LOG.warn("File could not be rotated");
    +//At this point there is nothing to do.  In all likelihood any filesystem operations will fail.
    +//The next tuple will almost certainly fail to write and/or sync, which force a rotation.  That
    +//will give rotateAndReset() a chance to work which includes creating a fresh file handle.
 }</li>
	<li>} catch (IOException e) 
{
    -    this.collector.reportError(e);
    -    this.collector.fail(tuple);
     }
<p> }</p></li>
</ul>


<p>    +    private void rotateAndReset() throws IOException {<br/>
    &#8212; End diff &#8211;</p>

<p>    What would be the behavior if the file system goes into an irrecoverable state where writes are continuously failing? Wont it end up creating a lot of empty files? You might want to consider cleaning this up and also killing the task by throwing a runtime exception after 'n' failures and hope that the new worker might be able to start writing successfully. </p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user dossett commented on a diff in the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/664#discussion_r37241329" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/664#discussion_r37241329</a></p>

<p>    &#8212; Diff: external/storm-hdfs/src/main/java/org/apache/storm/hdfs/bolt/HdfsBolt.java &#8212;<br/>
    @@ -80,6 +86,11 @@ public HdfsBolt addRotationAction(RotationAction action)</p>
{
     return this;
 }

<p>    +    public HdfsBolt withTickTupleIntervalSeconds(int interval) {<br/>
    &#8212; End diff &#8211;</p>

<p>    I like that change since it separates the function from the implementation.  A similar change could be made to HiveBolt, to which I also added a tick tuple functionality recently.  If this PR is otherwise ok, I'd like to open a new JIRA to change this bolt and HiveBolt at the same time.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user dossett commented on a diff in the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/664#discussion_r37241514" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/664#discussion_r37241514</a></p>

<p>    &#8212; Diff: external/storm-hdfs/src/main/java/org/apache/storm/hdfs/bolt/HdfsBolt.java &#8212;<br/>
    @@ -88,35 +99,94 @@ public void doPrepare(Map conf, TopologyContext topologyContext, OutputCollector</p>

<p> @Override<br/>
 public void execute(Tuple tuple) {</p>
<ul class="alternate" type="square">
	<li>try {</li>
	<li>byte[] bytes = this.format.format(tuple);</li>
	<li>synchronized (this.writeLock) {</li>
	<li>out.write(bytes);</li>
	<li>this.offset += bytes.length;<br/>
    -</li>
	<li>if (this.syncPolicy.mark(tuple, this.offset)) {</li>
	<li>if (this.out instanceof HdfsDataOutputStream) 
{
    -((HdfsDataOutputStream) this.out).hsync(EnumSet.of(SyncFlag.UPDATE_LENGTH));
    -    }
<p> else </p>
{
    -this.out.hsync();
    -    }</li>
	<li>this.syncPolicy.reset();<br/>
    +boolean forceRotate = false;<br/>
    +synchronized (this.writeLock) {<br/>
    +    boolean forceSync = false;<br/>
    +    if (TupleUtils.isTick(tuple)) 
{
    +LOG.debug("TICK! forcing a file system flush");
    +forceSync = true;
    +    }
<p>    +    else </p>
<div class="error"><span class="error">Unknown macro: {    +try {
    +    writeAndAddTuple(tuple);
    +} catch (IOException e) {
    +    //If the write failed, try to sync anything already written
    +    LOG.info("Tuple failed to write, forcing a flush of existing data.");
    +    this.collector.reportError(e);
    +    forceSync = true;
    +    this.collector.fail(tuple);
     } }</span> </div></li>
</ul>


<ul class="alternate" type="square">
	<li>this.collector.ack(tuple);<br/>
    +    if (this.syncPolicy.mark(tuple, this.offset) || forceSync) 
<div class="error"><span class="error">Unknown macro: {    +try {
    +    syncAndAckTuples();
    +} catch (IOException e) {
    +    LOG.warn("Data could not be synced to filesystem, failing this batch of tuples");
    +    this.collector.reportError(e);
    +    //Force rotation to get a new file handle
    +    forceRotate = true;
    +    for (Tuple t : tupleBatch)
    +this.collector.fail(t);
    +    tupleBatch.clear();
    +}    +    }</span> </div>
<p>    +}</p></li>
</ul>


<ul class="alternate" type="square">
	<li>if(this.rotationPolicy.mark(tuple, this.offset)){</li>
	<li>rotateOutputFile(); // synchronized</li>
	<li>this.offset = 0;</li>
	<li>this.rotationPolicy.reset();<br/>
    +if(this.rotationPolicy.mark(tuple, this.offset) || forceRotate) {<br/>
    +    try 
{
    +rotateAndReset();
    +    }
<p> catch (IOException e) </p>
{
    +this.collector.reportError(e);
    +LOG.warn("File could not be rotated");
    +//At this point there is nothing to do.  In all likelihood any filesystem operations will fail.
    +//The next tuple will almost certainly fail to write and/or sync, which force a rotation.  That
    +//will give rotateAndReset() a chance to work which includes creating a fresh file handle.
 }</li>
	<li>} catch (IOException e) 
{
    -    this.collector.reportError(e);
    -    this.collector.fail(tuple);
     }
<p> }</p></li>
</ul>


<p>    +    private void rotateAndReset() throws IOException {<br/>
    &#8212; End diff &#8211;</p>

<p>    I suppose it's possible that empty files would be created, but if a filesystem is in such a bad state I would expect creating new files to fail as well.</p>

<p>    Killing the task after enough failures is an interesting idea... what default value for # of attempts would you consider reasonable?</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user dossett commented on a diff in the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/664#discussion_r37242296" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/664#discussion_r37242296</a></p>

<p>    &#8212; Diff: external/storm-hdfs/src/main/java/org/apache/storm/hdfs/bolt/HdfsBolt.java &#8212;<br/>
    @@ -88,35 +99,94 @@ public void doPrepare(Map conf, TopologyContext topologyContext, OutputCollector</p>

<p> @Override<br/>
 public void execute(Tuple tuple) {</p>
<ul class="alternate" type="square">
	<li>try {</li>
	<li>byte[] bytes = this.format.format(tuple);</li>
	<li>synchronized (this.writeLock) {</li>
	<li>out.write(bytes);</li>
	<li>this.offset += bytes.length;<br/>
    -</li>
	<li>if (this.syncPolicy.mark(tuple, this.offset)) {</li>
	<li>if (this.out instanceof HdfsDataOutputStream) 
{
    -((HdfsDataOutputStream) this.out).hsync(EnumSet.of(SyncFlag.UPDATE_LENGTH));
    -    }
<p> else </p>
{
    -this.out.hsync();
    -    }</li>
	<li>this.syncPolicy.reset();<br/>
    +boolean forceRotate = false;<br/>
    +synchronized (this.writeLock) {<br/>
    +    boolean forceSync = false;<br/>
    +    if (TupleUtils.isTick(tuple)) 
{
    +LOG.debug("TICK! forcing a file system flush");
    +forceSync = true;
    +    }
<p>    +    else </p>
<div class="error"><span class="error">Unknown macro: {    +try {
    +    writeAndAddTuple(tuple);
    +} catch (IOException e) {
    +    //If the write failed, try to sync anything already written
    +    LOG.info("Tuple failed to write, forcing a flush of existing data.");
    +    this.collector.reportError(e);
    +    forceSync = true;
    +    this.collector.fail(tuple);
     } }</span> </div></li>
</ul>


<ul class="alternate" type="square">
	<li>this.collector.ack(tuple);<br/>
    +    if (this.syncPolicy.mark(tuple, this.offset) || forceSync) 
<div class="error"><span class="error">Unknown macro: {    +try {
    +    syncAndAckTuples();
    +} catch (IOException e) {
    +    LOG.warn("Data could not be synced to filesystem, failing this batch of tuples");
    +    this.collector.reportError(e);
    +    //Force rotation to get a new file handle
    +    forceRotate = true;
    +    for (Tuple t : tupleBatch)
    +this.collector.fail(t);
    +    tupleBatch.clear();
    +}    +    }</span> </div>
<p>    +}</p></li>
</ul>


<ul class="alternate" type="square">
	<li>if(this.rotationPolicy.mark(tuple, this.offset)){</li>
	<li>rotateOutputFile(); // synchronized</li>
	<li>this.offset = 0;</li>
	<li>this.rotationPolicy.reset();<br/>
    +if(this.rotationPolicy.mark(tuple, this.offset) || forceRotate) {<br/>
    +    try 
{
    +rotateAndReset();
    +    }
<p> catch (IOException e) </p>
{
    +this.collector.reportError(e);
    +LOG.warn("File could not be rotated");
    +//At this point there is nothing to do.  In all likelihood any filesystem operations will fail.
    +//The next tuple will almost certainly fail to write and/or sync, which force a rotation.  That
    +//will give rotateAndReset() a chance to work which includes creating a fresh file handle.
 }</li>
	<li>} catch (IOException e) 
{
    -    this.collector.reportError(e);
    -    this.collector.fail(tuple);
     }
<p> }</p></li>
</ul>


<p>    +    private void rotateAndReset() throws IOException </p>
{
    +rotateOutputFile(); // synchronized
    +this.offset = 0;
    +this.rotationPolicy.reset();
    +    }
<p>    +<br/>
    +    private void syncAndAckTuples() throws IOException {<br/>
    &#8212; End diff &#8211;</p>

<p>    Agreed, just pushed a commit to do that.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user arunmahadevan commented on a diff in the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/664#discussion_r37263240" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/664#discussion_r37263240</a></p>

<p>    &#8212; Diff: external/storm-hdfs/src/main/java/org/apache/storm/hdfs/bolt/HdfsBolt.java &#8212;<br/>
    @@ -80,6 +86,11 @@ public HdfsBolt addRotationAction(RotationAction action)</p>
{
     return this;
 }

<p>    +    public HdfsBolt withTickTupleIntervalSeconds(int interval) {<br/>
    &#8212; End diff &#8211;</p>

<p>    Assuming that would address default, min and max values for the interval, should be fine.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user arunmahadevan commented on a diff in the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/664#discussion_r37263411" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/664#discussion_r37263411</a></p>

<p>    &#8212; Diff: external/storm-hdfs/src/main/java/org/apache/storm/hdfs/bolt/HdfsBolt.java &#8212;<br/>
    @@ -88,35 +99,94 @@ public void doPrepare(Map conf, TopologyContext topologyContext, OutputCollector</p>

<p> @Override<br/>
 public void execute(Tuple tuple) {</p>
<ul class="alternate" type="square">
	<li>try {</li>
	<li>byte[] bytes = this.format.format(tuple);</li>
	<li>synchronized (this.writeLock) {</li>
	<li>out.write(bytes);</li>
	<li>this.offset += bytes.length;<br/>
    -</li>
	<li>if (this.syncPolicy.mark(tuple, this.offset)) {</li>
	<li>if (this.out instanceof HdfsDataOutputStream) 
{
    -((HdfsDataOutputStream) this.out).hsync(EnumSet.of(SyncFlag.UPDATE_LENGTH));
    -    }
<p> else </p>
{
    -this.out.hsync();
    -    }</li>
	<li>this.syncPolicy.reset();<br/>
    +boolean forceRotate = false;<br/>
    +synchronized (this.writeLock) {<br/>
    +    boolean forceSync = false;<br/>
    +    if (TupleUtils.isTick(tuple)) 
{
    +LOG.debug("TICK! forcing a file system flush");
    +forceSync = true;
    +    }
<p>    +    else </p>
<div class="error"><span class="error">Unknown macro: {    +try {
    +    writeAndAddTuple(tuple);
    +} catch (IOException e) {
    +    //If the write failed, try to sync anything already written
    +    LOG.info("Tuple failed to write, forcing a flush of existing data.");
    +    this.collector.reportError(e);
    +    forceSync = true;
    +    this.collector.fail(tuple);
     } }</span> </div></li>
</ul>


<ul class="alternate" type="square">
	<li>this.collector.ack(tuple);<br/>
    +    if (this.syncPolicy.mark(tuple, this.offset) || forceSync) 
<div class="error"><span class="error">Unknown macro: {    +try {
    +    syncAndAckTuples();
    +} catch (IOException e) {
    +    LOG.warn("Data could not be synced to filesystem, failing this batch of tuples");
    +    this.collector.reportError(e);
    +    //Force rotation to get a new file handle
    +    forceRotate = true;
    +    for (Tuple t : tupleBatch)
    +this.collector.fail(t);
    +    tupleBatch.clear();
    +}    +    }</span> </div>
<p>    +}</p></li>
</ul>


<ul class="alternate" type="square">
	<li>if(this.rotationPolicy.mark(tuple, this.offset)){</li>
	<li>rotateOutputFile(); // synchronized</li>
	<li>this.offset = 0;</li>
	<li>this.rotationPolicy.reset();<br/>
    +if(this.rotationPolicy.mark(tuple, this.offset) || forceRotate) {<br/>
    +    try 
{
    +rotateAndReset();
    +    }
<p> catch (IOException e) </p>
{
    +this.collector.reportError(e);
    +LOG.warn("File could not be rotated");
    +//At this point there is nothing to do.  In all likelihood any filesystem operations will fail.
    +//The next tuple will almost certainly fail to write and/or sync, which force a rotation.  That
    +//will give rotateAndReset() a chance to work which includes creating a fresh file handle.
 }</li>
	<li>} catch (IOException e) 
{
    -    this.collector.reportError(e);
    -    this.collector.fail(tuple);
     }
<p> }</p></li>
</ul>


<p>    +    private void rotateAndReset() throws IOException {<br/>
    &#8212; End diff &#8211;</p>

<p>    You could have a low default (say 10 continuous failures or so) and also add an option for the users to change this. @harshach can you recommend a default number of retries after which the task can exit so that storm would spawn the task in another worker where it might succeed?</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user harshach commented on a diff in the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/664#discussion_r38272576" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/664#discussion_r38272576</a></p>

<p>    &#8212; Diff: external/storm-hdfs/src/main/java/org/apache/storm/hdfs/bolt/HdfsBolt.java &#8212;<br/>
    @@ -88,35 +99,94 @@ public void doPrepare(Map conf, TopologyContext topologyContext, OutputCollector</p>

<p> @Override<br/>
 public void execute(Tuple tuple) {</p>
<ul class="alternate" type="square">
	<li>try {</li>
	<li>byte[] bytes = this.format.format(tuple);</li>
	<li>synchronized (this.writeLock) {</li>
	<li>out.write(bytes);</li>
	<li>this.offset += bytes.length;<br/>
    -</li>
	<li>if (this.syncPolicy.mark(tuple, this.offset)) {</li>
	<li>if (this.out instanceof HdfsDataOutputStream) 
{
    -((HdfsDataOutputStream) this.out).hsync(EnumSet.of(SyncFlag.UPDATE_LENGTH));
    -    }
<p> else </p>
{
    -this.out.hsync();
    -    }</li>
	<li>this.syncPolicy.reset();<br/>
    +boolean forceRotate = false;<br/>
    +synchronized (this.writeLock) {<br/>
    +    boolean forceSync = false;<br/>
    +    if (TupleUtils.isTick(tuple)) 
{
    +LOG.debug("TICK! forcing a file system flush");
    +forceSync = true;
    +    }
<p>    +    else </p>
<div class="error"><span class="error">Unknown macro: {    +try {
    +    writeAndAddTuple(tuple);
    +} catch (IOException e) {
    +    //If the write failed, try to sync anything already written
    +    LOG.info("Tuple failed to write, forcing a flush of existing data.");
    +    this.collector.reportError(e);
    +    forceSync = true;
    +    this.collector.fail(tuple);
     } }</span> </div></li>
</ul>


<ul class="alternate" type="square">
	<li>this.collector.ack(tuple);<br/>
    +    if (this.syncPolicy.mark(tuple, this.offset) || forceSync) 
<div class="error"><span class="error">Unknown macro: {    +try {
    +    syncAndAckTuples();
    +} catch (IOException e) {
    +    LOG.warn("Data could not be synced to filesystem, failing this batch of tuples");
    +    this.collector.reportError(e);
    +    //Force rotation to get a new file handle
    +    forceRotate = true;
    +    for (Tuple t : tupleBatch)
    +this.collector.fail(t);
    +    tupleBatch.clear();
    +}    +    }</span> </div>
<p>    +}</p></li>
</ul>


<ul class="alternate" type="square">
	<li>if(this.rotationPolicy.mark(tuple, this.offset)){</li>
	<li>rotateOutputFile(); // synchronized</li>
	<li>this.offset = 0;</li>
	<li>this.rotationPolicy.reset();<br/>
    +if(this.rotationPolicy.mark(tuple, this.offset) || forceRotate) {<br/>
    +    try 
{
    +rotateAndReset();
    +    }
<p> catch (IOException e) </p>
{
    +this.collector.reportError(e);
    +LOG.warn("File could not be rotated");
    +//At this point there is nothing to do.  In all likelihood any filesystem operations will fail.
    +//The next tuple will almost certainly fail to write and/or sync, which force a rotation.  That
    +//will give rotateAndReset() a chance to work which includes creating a fresh file handle.
 }</li>
	<li>} catch (IOException e) 
{
    -    this.collector.reportError(e);
    -    this.collector.fail(tuple);
     }
<p> }</p></li>
</ul>


<p>    +    private void rotateAndReset() throws IOException {<br/>
    &#8212; End diff &#8211;</p>

<p>    @dossett <br/>
    "I suppose it's possible that empty files would be created, but if a filesystem is in such a bad state I would expect creating new files to fail as well."<br/>
    This might bring down namenode <img class="emoticon" src="https://issues.apache.org/jira/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/>  with too many files.</p>

<p>    as @arunmahadevan pointed we should have retryCount I would start with 3 failed attempts and a configurable option</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user dossett commented on the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/664#issuecomment-136433305" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/664#issuecomment-136433305</a></p>

<p>    Thank you @harshach and @arunmahadevan, I will incorporate this feedback soon.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user dossett commented on the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/664#issuecomment-136499556" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/664#issuecomment-136499556</a></p>

<p>    @harshach @arunmahadevan Changes made based on feedback, PR also rebased and squashed to one commit.  Additional feedback welcome, thank you!</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user arunmahadevan commented on the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/664#issuecomment-136775368" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/664#issuecomment-136775368</a></p>

<p>    +1 overall the patch looks good once the <span class="error">&#91;comment&#93;</span> (<a href="https://github.com/apache/storm/pull/664#commitcomment-13003095" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/664#commitcomment-13003095</a>) is addressed</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user dossett commented on the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/664#issuecomment-136876170" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/664#issuecomment-136876170</a></p>

<p>    @arunmahadevan Our approach is to set the tick tuple frequency to be half of the message timeout setting for the topology.  Can the bolt get access to that topology setting? prepare() passes a TopologyContext to the bolt but I don't see a way to get configuration information out of it.  However, I am not that familiar with TopologyContext.</p>

<p>    If that's not possible, the bolt could just set a flush frequency of 15 seconds since the storm timeout default is 30 seconds and assume that anyone changing the timeout setting should also be adjusting the flush frequency as well</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user arunmahadevan commented on the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/664#issuecomment-136889084" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/664#issuecomment-136889084</a></p>

<p>    The value of `topology.message.timeout.secs` can be read from the `conf` that gets passed as the first parameter to `doPrepare`</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user dossett commented on the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/664#issuecomment-137064748" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/664#issuecomment-137064748</a></p>

<p>    @arunmahadevan So it can!  PR updated.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user dossett commented on the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/664#issuecomment-138607186" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/664#issuecomment-138607186</a></p>

<p>    @arunmahadevan @harshach Do you have any further comments about my updated PR? Thanks!</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user arunmahadevan commented on the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/664#issuecomment-138639789" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/664#issuecomment-138639789</a></p>

<p>    Looks good to me.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user dossett commented on the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/664#issuecomment-143341056" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/664#issuecomment-143341056</a></p>

<p>    @harshach (or other commiters) do you have feedback about this PR?</p>

<p>    Anecdotally, this has been very useful to us in production. We had an HDFS restart, which created the exact situation I tested with (failed writes due to HDFS safe mode) but the bolt recovered without a topology restart.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user harshach commented on the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/664#issuecomment-143884369" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/664#issuecomment-143884369</a></p>

<p>    +1</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user harshach commented on the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/664#issuecomment-143894338" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/664#issuecomment-143894338</a></p>

<p>    Thanks @dossett merged into master.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user asfgit closed the pull request at:</p>

<p>    <a href="https://github.com/apache/storm/pull/664" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/664</a></p>

**Comment by jira-importer** [comment]
<i>[Duplicates: <a href="https://github.com/apache/storm/issues?q=in%3Atitle%20STORM-1162">STORM-1162</a>]</i>

**Comment by jira-importer** [comment]
<i>[Originally duplicated by: <a href="https://github.com/apache/storm/issues?q=in%3Atitle%20STORM-804">STORM-804</a>]</i>

**Comment by jira-importer** [comment]
<i>[Originally related to: <a href="https://github.com/apache/storm/issues?q=in%3Atitle%20STORM-1073">STORM-1073</a>]</i>
